### DISCUSSION TITLE
What is the system default buffer size for MPI_Send? Are there any environment variables to set?

### DISCUSSION THREAD

**Discussion by duqi-hnu** [pmodels/mpich#7093] [Q&A]
What is the system default buffer size for MPI_Send? Are there any environment variables to set?

**Comment by hzhou** [comment]
Cound you provide some context about your question? Depending on the actual message paths and lower layer devices, different types and amounts of buffers may be used or may not use buffers at all.

**Reply by duqi-hnu** [reply]
Using ARM64 platform, using IB card, running FVCOM application, when the number of processes is less than or equal to 64, it can run normally, more than 64 on the abnormal exit.
The command that works is: FI_PROVIDER=verbs mpirun -n 64./fvcom
The command that doesn't work is: FI_PROVIDER=verbs mpirun -n 80. /fvcom
After debugging, the problem is found in the following code in fvcom, where NPROCS is the number of processes,
//------------
EC(:)%NSND = 0

DO I=1,NPROCS
IF(MYID /=I)THEN
ISND = MYID + 1000*I
CALL MPI_SEND(EC(I)%NRCV,1,MPI_INTEGER,I-1,ISND,MPI_COMM_WORLD,IERR)
END IF
END DO

DO I=1,NPROCS
IF(MYID /=I)THEN
IRCV = 1000*MYID + I
CALL MPI_RECV(EC(I)%NSND,1,MPI_INTEGER,I-1,IRCV,MPI_COMM_WORLD,STAT,IERR)
END IF
IF(EC(I)%NSND > 0)ALLOCATE(EC(I)%SNDP(EC(I)%NSND))
IF(EC(I)%NSND == 0)ALLOCATE(EC(I)%SNDP(1))
END DO
//----------------------
After debug, the problem is found in the following code in fvcom, where NPROCS is the number of processes,
After MPI_SEND has finished executing, MPI_RECV shows that the tag does not match and throws an error. This problem can be solved by placing MPI_RECV in the same code section as MPI_SEND, so it is reasonable to suspect that the size of the system's MPI Buffer is causing the error.
How can we fix this without modifying the application source code? @hzhou

**Reply by duqi-hnu** [reply]
MPICH-4.2.2

**Reply by hzhou** [reply]
Could you try changing the tag into `src_rank + 100 * dst_rank`?

**Reply by duqi-hnu** [reply]
After the tag is modified,  runs properly

**Reply by hzhou** [reply]
The issue is due to exceeding maximum tag value. You can query the upper bound of tag value using `MPI_Comm_get_attr` with key `MPI_TAG_UB`.

**Reply by duqi-hnu** [reply]
thanks
